jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -107,8 +107,8 @@ object XGBoost extends Serializable {
     // to workaround the empty partitions in training dataset,
     // this might not be the best efficient implementation, see
     // (https://github.com/dmlc/xgboost/issues/1277)
-    partitionedTrainingSet.mapPartitions {
-      trainingSamples =>
+    partitionedTrainingSet.mapPartitionsWithIndex {
+      case (partIndex, trainingSamples) =>
```

Comment:
I just noticed this change, you don't need to use mapPartitionsWithIndex, and you can just directly use `TaskContext.getPartitionId()` to get partitionIndex